// ember-vcl-plotly/demo/example2/application.js
import Ember from 'ember';

export default Ember.Controller.extend({
  point: null,
  actions: {
    plotClick({ points }) {
      const xys = points.map(point => `label: ${point.label} | value: ${point.text}\n`);
      this.set('point', xys);
    }
  },

  plotData: [{
    values: [16, 15, 12, 6, 5, 4, 42],
    labels: ['US', 'China', 'European Union', 'Russian Federation', 'Brazil', 'India', 'Rest of World' ],
    domain: {
      x: [0, .48]
    },
    name: 'GHG Emissions',
    hoverinfo: 'label+percent+name',
    hole: .4,
    type: 'pie'
  },{
    values: [27, 11, 25, 8, 1, 3, 25],
    labels: ['US', 'China', 'European Union', 'Russian Federation', 'Brazil', 'India', 'Rest of World' ],
    text: 'CO2',
    textposition: 'inside',
    domain: {x: [.52, 1]},
    name: 'CO2 Emissions',
    hoverinfo: 'label+percent+name',
    hole: .4,
    type: 'pie'
  }],

  plotLayout: {
    title: 'Global Emissions 1990-2011',
    width: 1200,
    height: 800,
    annotations: [
      {
        font: {
          size: 20
        },
        showarrow: false,
        text: 'GHG',
        x: 0.17,
        y: 0.5
      },
      {
        font: {
          size: 20
        },
        showarrow: false,
        text: 'CO2',
        x: 0.82,
        y: 0.5
      }
    ]
  }
});
